apps/api_web/lib/api_web/rate_limiter/rate_limiter_concurrent.ex

```diff
@@ -0,0 +1,175 @@
+defmodule ApiWeb.RateLimiter.RateLimiterConcurrent do
+  @moduledoc """
+  Rate limits a user's concurrent connections based on their API key or by their IP address if no
+  API key is provided. Split by static and event-stream requests.
+  """
+
+  use GenServer
+  require Logger
+
+  def start_link([]), do: GenServer.start_link(__MODULE__, nil, name: __MODULE__)
+
+  def init(_) do
+    connection_opts =
+      Keyword.fetch!(Application.get_env(:api_web, :rate_limiter_concurrent), :connection_opts)
+
+    {:ok, pid} = if memcache?(), do: Memcache.start_link(connection_opts), else: {:ok, nil}
+    {:ok, %{memcache_pid: pid}}
+  end
+
+  defp lookup(%ApiWeb.User{} = user, event_stream?) do
+    type = if event_stream?, do: "event_stream", else: "static"
+    key = "concurrent_#{user.id}_#{type}"
+
+    {
+      type,
+      key
+    }
+  end
+
+  defp get_current_unix_ts do
+    DateTime.utc_now() |> DateTime.to_unix()
+  end
+
+  defp get_heartbeat_tolerance do
+    Keyword.fetch!(Application.get_env(:api_web, :rate_limiter_concurrent), :heartbeat_tolerance)
+  end
+
+  def get_locks(%ApiWeb.User{} = user, event_stream?) do
+    if enabled?() do
+      current_timestamp = get_current_unix_ts()
+      heartbeat_tolerance = get_heartbeat_tolerance()
+      {_type, key} = lookup(user, event_stream?)
+      {:ok, locks} = GenServer.call(__MODULE__, {:memcache_get, key, %{}})
```

Comment:
suggestion: I think that having a `GenServer.call/2` here is going to mean we won't be able to have concurrent API calls: they'll be serialized as messages to this single GenServer. Could we share the pooled Memcache connections we're already using for the regular rate limiter?

Reply:
Are you saying that calling `GenServer.call` makes the entire call chain synchronous, or just the writes to memcache? 

I'm not entirely sure we would want asynchronous writes to memcache in this particular case due to the fact that it doesn't support transactions - this might inadvertently prevent the race mentioned earlier at least on a per-instance basis 🤔 

I have some test code locally that we may be able to use to take a similar approach w/r/t memcache supervisor if this doesn't work as expected.

Reply:
Thank you, that makes sense! I have updated to share the pool with `RateLimiter`.